### ISSUE TITLE
"Kitchen Converge" works but "Kitchen Test" fails on Converge step

### ISSUE THREAD

**Issue by atifrasheed79** [test-kitchen/test-kitchen#1515]
## Description

"Kitchen Converge" works but "Kitchen Test" fails on Converge step while it is going through bootstrap.sh where I am trying to install few gems.

```
#!/bin/bash
export BUSSER_ROOT="/tmp/verifier"
export GEM_HOME="/tmp/verifier/gems"
export GEM_PATH="/tmp/verifier/gems"
export GEM_CACHE="/tmp/verifier/gems/cache"
export PATH="$PATH:/tmp/verifier/bin:/tmp/verifier/gems/bin"
_gem install rdoc -v 4.3.0_
gem install bundler -v 1.17.3
gem install net-ssh -v 4.2
gem install busser-serverspec
gem install serverspec
gem install busser
```
It fails when it is trying to install rdoc and I get below error. 
```
Fetching: rdoc-4.3.0.gem (100%)
       Successfully installed rdoc-4.3.0
       ERROR:  While executing gem ... (NoMethodError)
           undefined method `reset' for RDoc::TopLevel:Class
```
As I mentioned earlier, it works fine when I do 1) kitchen create, 2) kitchen converge

## Kitchen Version
Test Kitchen version 1.16.0

## Ruby Version
ruby 2.0.0p648 (2015-12-16) [x86_64-linux]

## Platform Version
Amazon Linux release 2 (Karoo)

## Kitchen Diagnose
I have removed some sensitive information.

```
$ kitchen diagnose --all
---
timestamp: '2019-01-23 03:44:16 UTC'
kitchen_version: 1.16.0
plugins:
  driver:
    Ec2:
      class: Kitchen::Driver::Ec2
      version: 1.3.2
      api_version: 2
  provisioner:
    Shell:
      class: Kitchen::Provisioner::Shell
      version: 1.16.0
      api_version: 2
  transport:
    Ssh:
      class: Kitchen::Transport::Ssh
      version: 1.16.0
      api_version: 1
  verifier:
    Busser:
      class: Kitchen::Verifier::Busser
      version: 1.16.0
      api_version: 1
loader:
  process_erb: true
  process_local: true
  process_global: true
  global_config:
  project_config:
    filename: /home/ec2-user/deploy/packer/tests/.kitchen.yml
    raw_data:
      driver:
        name: ec2
        region: ap-southeast-2
        instance_type: t2.small
        associate_public_ip: false
      transport:
        connection_timeout: 10
        connection_retries: 5
        username: ec2-user
      busser:
        ruby_bindir: /usr/bin
      provisioner:
        name: shell
      platforms:
        driver:
      suites:
      - name: default
  local_config:
  combined_config:
    filename:
    raw_data:
      driver:
        name: ec2
        instance_type: t2.small
        associate_public_ip: false
      transport:
        connection_timeout: 10
        connection_retries: 5
        username: ec2-user
      busser:
        ruby_bindir: /usr/bin
      provisioner:
        name: shell
      platforms:
        driver:
      suites:
      - name: default
instances:
    platform:
      os_type: unix
      shell_type: bourne
    state_file:
      last_action: create
      last_error:
    driver:
      associate_public_ip: false
      availability_zone:
      aws_access_key_id:
      aws_secret_access_key:
      aws_session_token:
      aws_ssh_key_id: codeninjas-jenkins
      block_duration_minutes:
      ebs_optimized: false
      http_proxy:
      iam_profile_name:
      image_search:
      instance_initiated_shutdown_behavior:
      instance_type: t2.small
      interface:
      kitchen_root: /home/ec2-user/deploy/packer/tests
      log_level: :info
      name: ec2
      private_ip_address:
      region: ap-southeast-2
      retry_limit: 3
      retryable_sleep: 5
      retryable_tries: 60
      tenancy: default
      test_base_path: /home/ec2-user/deploy/packer/tests/test/integration
      user_data:
      username:
    provisioner:
      command_prefix:
      data_path:
      ftp_proxy:
      http_proxy:
      https_proxy:
      kitchen_root: /home/ec2-user/deploy/packer/tests
      max_retries: 1
      name: shell
      retry_on_exit_code: []
      root_path: /tmp/kitchen
      script: /home/ec2-user/deploy/packer/tests/bootstrap.sh
      sudo: true
      sudo_command: sudo -E
      test_base_path: /home/ec2-user/deploy/packer/tests/test/integration
      wait_for_retry: 30
    transport:
      compression: false
      compression_level: 0
      connection_retries: 5
      connection_retry_sleep: 1
      connection_timeout: 10
      keepalive: true
      keepalive_interval: 60
      kitchen_root: /home/ec2-user/deploy/packer/tests
      log_level: :info
      max_ssh_sessions: 9
      max_wait_until_ready: 600
      name: ssh
      port: 22
      ssh_gateway:
      ssh_gateway_username:
      test_base_path: /home/ec2-user/deploy/packer/tests/test/integration
      username: ec2-user
    verifier:
      busser_bin: /tmp/verifier/bin/busser
      chef_omnibus_root: /opt/chef
      command_prefix:
      ftp_proxy:
      http_proxy:
      https_proxy:
      kitchen_root: /home/ec2-user/deploy/packer/tests
      log_level: :info
      name: busser
      root_path: /tmp/verifier
      ruby_bindir: /usr/bin
      sudo: true
      sudo_command: sudo -E
      suite_name: default
      test_base_path: /home/ec2-user/deploy/packer/tests/test/integration
      version: busser
```

**Comment by atifrasheed79** [comment]
sorry, just realized, it's the second converge which works, 1st one fails. Probably need to setup the gem env in userdata.

**Comment by cheeseplus** [comment]
On the surface this seems less like a kitchen specific issue and more "provisioner semantics" along with some rather old/deprecated tk/Ruby. If you could share a redacted `kitchen.yml` to make reproducing this easier, that'd be great.

**Comment by antonu17** [comment]
Hi there.
Happy to share I'm facing exactly the same behavior.
When I run `kitchen destroy`, `kitchen create` and `kitchen converge` it works. But `kitchen test` fails on converge step.
I'm using `ec2` driver and `ansible_push` provisioner.
ruby 2.5.1p57
test-kitchen 1.24.0
kitchen-ansiblepush 0.9.1
In my case this simple playbook fails:
```
- name: ensure python installed
  hosts: all
  gather_facts: False
  tasks:
    - name: install python 2
      raw: test -e /usr/bin/python || (apt -y update && apt install -y python-minimal)
      changed_when: False
```

**Comment by atifrasheed79** [comment]
Hello everyone, just to update I was able to work it around by setting the BUSSER_HOME, GEM_HOME, GEM_PATH etc. in UserData (kitchen.yml) for the EC2 instance Kitchen will launch, and bootstrap.sh. 

When I enabled debug, I found that Kitchen converge and setup are working in a different environment when I was doing "Kitchen test", so I just made sure both steps refer to same GEM environment. 

So just enable debug when you are doing "Kitchen Test" to make sure all Kitchen Test phases are working in same GEM environment.

kitchen test --log-level=debug
